### PR TITLE
Add 65C02 STZ instructions.

### PR DESCRIPTION
**Pull request by jackoalan**
Zeros a memory location using one of four address modes. If the instruction selection for G_STORE is able to determine if a zero constant is stored, the best matching STZ is emitted.

### CODE REVIEW

File: llvm/lib/Target/MOS/MOSInstrLogical.td [hunk 1]
```diff
@@ -355,6 +355,12 @@ def STYIndir : MOSStore, PseudoInstExpansion<(STA_IndirectIndexed addr8:$addr)>
   dag InOperandList = (ins Ac:$src, Imag16:$addr, Yc:$offset);
 }
 
+// STZ abs,x
+def STZIdx : MOSStore {
```
[the remark this answers is not in the document]
Comment (johnwbyrd) [reply llvm/lib/Target/MOS/MOSInstrLogical.td:359]: The low-level assembly emission looks correct per http://6502.org/tutorials/65c02opcodes.html.  I know that mysterymath wants the code lowered slightly differently.

[the remark this answers is not in the document]
Comment (johnwbyrd) [reply llvm/lib/Target/MOS/MOSInstrLogical.td:359]: Good to have you working on this, by the way, jackoalan.

[the remark this answers is not in the document]
Comment (jackoalan) [reply llvm/lib/Target/MOS/MOSInstrLogical.td]: Done! `PseudoInstExpansion` is really cool

File: llvm/lib/Target/MOS/MOSInstructionSelector.cpp [hunk 2]
```diff
@@ -606,15 +612,22 @@ bool MOSInstructionSelector::selectLoadStore(MachineInstr &MI) {
     AbsOpcode = MOS::STAbs;
     IdxOpcode = MOS::STIdx;
     YIndirOpcode = MOS::STYIndir;
+    // STZ
+    if (STI.has65C02() && isOperandImmEqual(SrcDstOp, 0, MRI)) {
+      AbsOpcode = MOS::STZ_Absolute;
```
[the remark this answers is not in the document]
Comment (jackoalan) [reply llvm/lib/Target/MOS/MOSInstructionSelector.cpp]: Makes sense, fixed

**Comment by mysterymath** [comment]
LGTM!

As an aside, I'd guess the biggest benefit we'd see from STZ wouldn't be direct patterns of STZ like this; instead it'd be the clearing of imaginary registers with the ZP version of STZ. This would need to be modelled differently: we'd probably want something like a LDZ pseudoinstruction that mirrors the LDImm instruction, but only on the 65C02. This could then lower to either LD{AXY} #0 or into an STZ zp, depending on what the output was (either GPR or Imag8). We'd then need to select/emit that wherever we emit LDImms.

This is a bit of additional work, and you may already have something along these lines, so I'm not blocking this on that. Thanks for the PR!

**Pull request merged**